File: programmers/알고리즘 레벨2/PG_118667_두_큐_합_같게_만들기.cpp
// Solve 2023-11-25

#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(vector<int> queue1, vector<int> queue2) {
    queue<int> que1;
    queue<int> que2;
    long sum_of_que1 = 0;
    long sum_of_que2 = 0;

    for (int x : queue1) {
        que1.push(x);
        sum_of_que1 += x;
    }

    for (int x : queue2) {
        que2.push(x);
        sum_of_que2 += x;
    }

    if ((sum_of_que1 + sum_of_que2) % 2 == 1) {
        return -1;
    }

    int iter = 0;
    int max_iter = (int)que1.size() * 4;

    while (iter < max_iter) {
        if (sum_of_que1 > sum_of_que2) {
            sum_of_que1 -= que1.front();
            que2.push(que1.front());
            que1.pop();
            sum_of_que2 += que2.back();
        }
        else if (sum_of_que1 < sum_of_que2) {
            sum_of_que2 -= que2.front();
            que1.push(que2.front());
            que2.pop();
            sum_of_que1 += que1.back();
        }
        else {
            return iter;
        }

        iter++;
    }

    return -1;
}

File: programmers/algo/110000+/PG_118667_두_큐_합_같게_만들기.java
// Solve 2025-05-21

import java.util.*;

class PG_118667_두_큐_합_같게_만들기 {

    public int solution(int[] queue1, int[] queue2) {
        ArrayDeque<Integer> que1 = new ArrayDeque<>();
        ArrayDeque<Integer> que2 = new ArrayDeque<>();
        long sumOfQue1 = 0;
        long sumOfQue2 = 0;

        for (int x : queue1) {
            que1.addLast(x);
            sumOfQue1 += x;
        }

        for (int x : queue2) {
            que2.addLast(x);
            sumOfQue2 += x;
        }

        if ((sumOfQue1 + sumOfQue2) % 2 == 1) {
            return -1;
        }

        int iter = 0;
        int maxIter = queue1.length * 4;

        while (iter < maxIter) {
            if (sumOfQue1 > sumOfQue2) {
                int cur = que1.pollFirst();
                que2.addLast(cur);
                sumOfQue1 -= cur;
                sumOfQue2 += cur;
            } else if (sumOfQue1 < sumOfQue2) {
                int cur = que2.pollFirst();
                que1.addLast(cur);
                sumOfQue2 -= cur;
                sumOfQue1 += cur;
            } else {
                return iter;
            }

            iter++;
        }

        return -1;
    }

}
